# Анализ книжного рынка

## Описание проекта

## Библиотеки и подключение к базе данных

In [1]:
# импортируем библиотеки 
import pandas as pd 
from sqlalchemy import create_engine 

In [2]:
# устанавливаем параметры 
db_config = {'user': 'praktikum_student', # имя пользователя 
                     'pwd': 'Sdf4$2;d-d30pp', # пароль 
                     'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net', 
                     'port': 6432, # порт подключения 
                     'db': 'data-analyst-final-project-db'} # название базы данных 

# Формируем строку соединения с БД
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'], 
                                                        db_config['pwd'], 
                                                        db_config['host'], 
                                                        db_config['port'], 
                                                        db_config['db']) 
# сохраняем коннектор 
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

## Исследуйте таблицы — выведите первые строки

### Таблица books

In [3]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                books
            ORDER BY 
                book_id
            LIMIT
                5
        '''

In [4]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                books
            ORDER BY 
                book_id DESC
            LIMIT
                5
        '''

In [6]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

book_id  author_id                                              title  \
0     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   
1      999         94                        You Suck (A Love Story  #2)   
2      998        201                                    Year of Wonders   
3      997        454                        Xenocide (Ender's Saga  #3)   
4      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   

   num_pages publication_date  publisher_id  
0        540       2006-04-25           143  
1        328       2007-01-16           331  
2        358       2002-04-30           212  
3        592       1996-07-15           297  
4        265       2001-02-06           147

In [7]:
# SQL запрос
# поищем пропуски

query = ''' SELECT 
                *
            FROM 
                books
            WHERE
                book_id IS NULL
        '''

In [8]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, author_id, title, num_pages, publication_date, publisher_id]
Index: []

In [9]:
# проверим на дубликаты
query = '''    SELECT
                    book_id,
                    COUNT(book_id)
                FROM
                    books
                -- группируем выборку по значению
                GROUP BY
                    book_id
                -- фильтруем выборку по количеству
                HAVING
                    COUNT(book_id) > 1
        
        '''

In [10]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [book_id, count]
Index: []

У нас есть данные о публикации 1000 книг. Далее я выводил сразу всю табличку, они не большие. Дубликаты и пропуски мы не нашли

### Таблица authors

In [11]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                authors
        '''

In [12]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [13]:
# SQL запрос
# поищем пропуски

query = ''' SELECT 
                *
            FROM 
                authors
            WHERE
                author_id IS NULL
        '''

In [14]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id, author]
Index: []

In [15]:
# проверим на дубликаты
query = '''    SELECT
                    author_id,
                    COUNT(author_id)
                FROM
                    authors
                -- группируем выборку по значению
                GROUP BY
                    author_id
                -- фильтруем выборку по количеству
                HAVING
                    COUNT(author_id) > 1
        
        '''

In [16]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [author_id, count]
Index: []

У нас 636 авторов. Пропусков и дубликатов нет

### Таблица publishers

In [17]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                publishers
        '''

In [18]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [19]:
# SQL запрос
# поищем пропуски

query = ''' SELECT 
                *
            FROM 
                publishers
            WHERE
                publisher_id IS NULL
        '''

In [20]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id, publisher]
Index: []

In [21]:
# проверим на дубликаты
query = '''    SELECT
                    publisher_id,
                    COUNT(publisher_id)
                FROM
                    publishers
                -- группируем выборку по значению
                GROUP BY
                    publisher_id
                -- фильтруем выборку по количеству
                HAVING
                    COUNT(publisher_id) > 1
        
        '''

In [22]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [publisher_id, count]
Index: []

Так же у нас информация о 340 издательствах с их названиями. Пропуски и дубликаты мы не нашли

### Таблица ratings

In [23]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                ratings
        '''

In [24]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [25]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                ratings
            WHERE
                rating_id IS NULL
        '''

In [26]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id, book_id, username, rating]
Index: []

In [27]:
# проверим на дубликаты
query = '''    SELECT
                    rating_id,
                    COUNT(rating_id)
                FROM
                    ratings
                -- группируем выборку по значению
                GROUP BY
                    rating_id
                -- фильтруем выборку по количеству
                HAVING
                    COUNT(rating_id) > 1
        
        '''

In [28]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [rating_id, count]
Index: []

Далее у нас информация о 6456 оценках от пользователей. Многие книги видимо были оценены и не раз, так же, как и многие пользователи оценили не одну книгу. Пропуски и дубликаты мы не нашли

### Таблица reviews

In [29]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                reviews
        '''

In [30]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

review_id  book_id        username  \
0             1        1    brandtandrea   
1             2        1      ryanfranco   
2             3        2        lorichen   
3             4        3   johnsonamanda   
4             5        3     scotttamara   
...         ...      ...             ...   
2788       2789      999      martinadam   
2789       2790     1000         wknight   
2790       2791     1000  carolrodriguez   
2791       2792     1000         wendy18   
2792       2793     1000      jarvispaul   

                                                   text  
0     Mention society tell send professor analysis. ...  
1     Foot glass pretty audience hit themselves. Amo...  
2     Listen treat keep worry. Miss husband tax but ...  
3     Finally month interesting blue could nature cu...  
4     Nation purpose heavy give wait song will. List...  
...                                                 ...  
2788  Later hospital turn easy community. Fact same ...  
2789   Change lose answer close pressure. Spend so now.  
2790  Authority go who television entire hair guy po...  
2791  Or western offer wonder ask. More hear phone f...  
2792  Republican staff bit eat material measure plan...  

[2793 rows x 4 columns]

In [31]:
# SQL запрос

query = ''' SELECT                 
                COUNT(DISTINCT book_id) AS cnt_book
            FROM 
                reviews
        '''

In [32]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

cnt_book
0       994

In [33]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                reviews
            WHERE
                book_id IS NULL
        '''

In [34]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

In [35]:
# SQL запрос

query = ''' SELECT 
                *
            FROM 
                reviews
            WHERE
                review_id IS NULL
        '''

In [36]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, book_id, username, text]
Index: []

In [37]:
# проверим на дубликаты
query = '''    SELECT
                    review_id,
                    COUNT(review_id)
                FROM
                    reviews
                -- группируем выборку по значению
                GROUP BY
                    review_id
                -- фильтруем выборку по количеству
                HAVING
                    COUNT(review_id) > 1
        
        '''

In [38]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

Empty DataFrame
Columns: [review_id, count]
Index: []

Ну а тут у нас информация о обзорах на книги от пользователей. Не для всех книг у нас есть обзоры

**Вывод**

Мы осмотрели все таблицы. В уникальных id мы не нашли пропуски или дубликатов. Не для всех книг у нас есть обзоры

<div class="alert alert-block alert-warning">
    

__Комментарий от ревьюера №1__

С нашими данными всё в порядке или есть какие-то проблемы?
</div>

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <font size="4"><b>Комментарии студента</b></font>
    
проверил на дубликаты и пропуски в уникальных индификаторах в каждой таблице 
</div>

## Посчитайте, сколько книг вышло после 1 января 2000 года

In [39]:
# SQL запрос

query = ''' SELECT 
                COUNT(book_id) AS book_cnt
            FROM 
                books
            WHERE
                publication_date > '2000-01-01'
        '''

In [40]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

book_cnt
0       819

После 01.01.2000 было выпущено 819 книг.

## Для каждой книги посчитайте количество обзоров и среднюю оценку

In [41]:
query = ''' SELECT 
               books.title AS title,
               COUNT(DISTINCT review_id) AS cnt_review,
               AVG(ratings.rating) AS mean_rating
            FROM
                books
                LEFT JOIN reviews ON books.book_id = reviews.book_id
                LEFT JOIN ratings ON books.book_id = ratings.book_id
            GROUP BY
                title
            ORDER BY 
                cnt_review DESC
        '''

In [42]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

title  cnt_review  \
0                                  Memoirs of a Geisha           8   
1                              Twilight (Twilight  #1)           7   
2    Harry Potter and the Chamber of Secrets (Harry...           6   
3                                     The Glass Castle           6   
4                                      Eat  Pray  Love           6   
..                                                 ...         ...   
994    Anne Rice's The Vampire Lestat: A Graphic Novel           0   
995                            The Natural Way to Draw           0   
996   The Cat in the Hat and Other Dr. Seuss Favorites           0   
997                          Essential Tales and Poems           0   
998  Disney's Beauty and the Beast (A Little Golden...           0   

     mean_rating  
0       4.138462  
1       3.662500  
2       4.287500  
3       4.206897  
4       3.395833  
..           ...  
994     3.666667  
995     3.000000  
996     5.000000  
997     4.000000  
998     4.000000  

[999 rows x 3 columns]

<div class="alert alert-block alert-warning">
    

__Комментарий от ревьюера №1__

В данном запросе какая-то проблема, как-то слишком много у нас ревью. Тут можно проверить сколько ревью было в изначальной таблице и сколько ревью получилось в агрегированной таблице.
</div>

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <font size="4"><b>Комментарии студента</b></font>
    
Да верно, появилось много дублирующих срок при соединение, поэтому поставил считать только уникальные обзоры, а не все строки
</div>

## Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры

In [43]:
query = ''' 

    SELECT
        MAX(num_books),
        publisher
    FROM (
        SELECT              
            COUNT(books.book_id) AS num_books,               
            publishers.publisher AS publisher
        FROM
            books
            LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
        WHERE
            num_pages > 50
        GROUP BY
            publisher
        ) AS SQUB 
    GROUP BY 
        publisher
    ORDER BY
        max DESC
    LIMIT 2
    

        '''

In [44]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

max      publisher
0   42  Penguin Books
1   31        Vintage

Пытался через MAX без сортировки, но никак не получается название издательства при этом вывести, только цифру

## Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками

In [45]:
query = '''        
        SELECT
            
            AVG(mean_rating) AS mean_total_rating,
            author
        FROM (    
            SELECT
                title,
                COUNT(rating) AS cnt_rating,
                AVG(rating) AS mean_rating,
                author
            FROM
                books
                LEFT JOIN ratings ON books.book_id = ratings.book_id
                LEFT JOIN authors ON books.author_id = authors.author_id
            GROUP BY
                title, author
            HAVING
                COUNT(rating) > 50
            ORDER BY
                author DESC                
            ) AS SQBU  
        GROUP BY
            author
        ORDER BY
            mean_total_rating DESC
        LIMIT 1    

        '''

In [46]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

mean_total_rating                      author
0           4.283844  J.K. Rowling/Mary GrandPré

СНачала мы сгруппировали по книгам и их авторам, посчитали кол-во оценок и их среднее. Далее получившийся результат сгруппировали уже только по авторам, и посчитав среднее вывели лидера: 	J.K. Rowling/Mary GrandPré с оценкой 4.283844

## Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [47]:
query = '''        
        SELECT
            AVG(cnt_review) AS mean_review
        FROM (
            SELECT
                COUNT(rating) AS cnt_rating,
                ratings.username,
                AVG(cnt_review) AS cnt_review
            FROM
                ratings
                LEFT JOIN (
                    SELECT
                        reviews.username,
                        COUNT(text) AS cnt_review
                    FROM
                        reviews    
                    GROUP BY
                        reviews.username
                    ) AS SQBU ON ratings.username = SQBU.username
            GROUP BY
                ratings.username
            HAVING
                COUNT(rating) > 50
            ) AS SQBU_2                          
                      
        '''

In [48]:
# Чтобы выполнить SQL-запрос, используйте Pandas
pd.io.sql.read_sql(query, con = engine)

mean_review
0    24.333333

В первом подзапросе мы сначала посчитали кол-во обзоров для каждого пользователя. Далее мы соединили с таблицей, где для каждого пользователя посчитано кол-во оценок и сделан срез более 50 и соединили таблицы. Т.к. к каждому пользователю при соединение таблиц будет добавляться кол-во сделанных им обзоров, то логично сделать среднее, а не считать строки. В конце мы это обрамляем в подзапрос и просто считаем среднее по столбцу кол-ва обзоров. В итоге получилась цифра 24.3

<div class="alert alert-block alert-success">
    

__Комментарий от ревьюера №1__

1, 3, 4 и 5 запросы сделаны верно, молодец.
</div>

## Вывод

- После 01.01.2000 было выпущено 819 книг.
- Издательство, выпустившее большее кол-во книг с кол-вом страниц более 50: Penguin Books, 42 книги
- Автор с самой высокой средней оценкой книг (более 50 оценок для книг): лидер J.K. Rowling/Mary GrandPré с оценкой 4.283844
- Среднее кол-во обзоров от пользователей, поставивших более 50 оценок:  24.3

Так же мы для каждой книги посчитали кол-во обзоров и среднюю оценку, тройка лидеров:

|title |	cnt_review |	mean_rating|
|------|---------------|---------------|
|Memoirs of a Geisha |	8 	|4.13 |
|Twilight (Twilight #1) |	7 |	3.66 |
|Harry Potter and the Chamber of Secrets (Harry... |	6 |	4.28 |


<div class="alert alert-block alert-warning">
    

__Комментарий от ревьюера №1__

Хорошо, что написал общий вывод, но его нужно будет поправить после исправления в проекте второго запроса)
</div>

<div class="alert alert-info" style="border-radius: 15px; box-shadow: 4px 4px 4px; border: 1px solid " > <font size="4"><b>Комментарии студента</b></font>
    
Поправил
</div>